src/java.base/share/classes/javax/crypto/KDFSpi.java

```diff
@@ -0,0 +1,153 @@
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public abstract class KDFSpi {
+
+    /**
+     * The sole constructor.
+     * <p>
+     * A {@code KDFParameters} object may be specified for KDF algorithms
+     * that support initialization parameters.
+     *
+     * @param kdfParameters
+     *     the initialization parameters for the {@code KDF} algorithm (may be
+     *     {@code null})
+     *
+     * @see KDF#getParameters()
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this
+     *     {@code KDFSpi}
+     */
+    protected KDFSpi(KDFParameters kdfParameters)
+        throws InvalidAlgorithmParameterException {}
+
+    /**
+     * Returns the {@code KDFParameters} used with this {@code KDF} object.
+     * <p>
+     * The returned parameters may be the same that were used to initialize
+     * this {@code KDF} object, or may contain additional default or
+     * random parameter values used by the underlying KDF algorithm.
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
+     *
+     * @return the parameters used with this {@code KDF} object, or
+     * {@code null}
+     */
+    protected abstract KDFParameters engineGetParameters();
+
+    /**
+     * Derives a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code engineDeriveKey} method may be called multiple times on a particular
+     * {@code KDFSpi} instance.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param derivationParameterSpec
+     *     derivation parameters
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the
+     *     KDF output and according to the derivation parameters.
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code derivationParameterSpec} is
+     *     invalid or if the combination of {@code alg} and the {@code derivationParameterSpec}
+     *     results in something invalid, ie - a key of inappropriate length
+     *     for the specified algorithm
+     * @throws NoSuchAlgorithmException
+     *     if {@code alg} is empty or invalid
+     * @throws NullPointerException
+     *     if {@code alg} or {@code derivationParameterSpec} is null
+     */
+    protected abstract SecretKey engineDeriveKey(String alg,
+                                                 AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException, NoSuchAlgorithmException;
+
+    /**
+     * Obtains raw data from a key derivation function.
+     * <p>
+     * The {@code engineDeriveData} method may be called multiple times on a
+     * particular {@code KDFSpi} instance.
```

Comment:
You will need to change the wording on this, and the `engineDeriveData` method, so as not to imply this supports concurrent access.